pypots/imputation/cdsa/modules/submodules.py

```diff
@@ -0,0 +1,243 @@
+
+        Parameters
+        ----------
+        q_time:
+            Query tensor on time-series.
+        k_time:
+            Key tensor on time-series.
+        q_feature:
+            Query tensor on feature.
+        k_feature:
+            Key tensor on feature.
+        v:
+            Value tensor.
+
+        attn_mask:
+            Masking tensor for the attention map. The shape should be [batch_size, n_heads, n_steps, n_steps].
+            0 in attn_mask means values at the according position in the attention map will be masked out.
+
+        Returns
+        -------
+        output:
+            The result of Value multiplied with the scaled dot-product attention map.
+
+        attn_time:
+            The scaled dot-product attention map on time-series.
+
+        attn_feature:
+            The scaled dot-product attention map on feature.
+        """
+        # q_time, k_time, v all have 4 dimensions [batch_size, n_heads, n_steps, d_tensor]
+        # d_tensor could be d_q, d_k, d_v
+
+        # dot product q with k.T to obtain similarity
+        attn_time = torch.matmul(
+            q_time / self.temperature, k_time.transpose(2, 3)
+        )  # [batch_size, n_heads, n_steps, n_steps]
+        attn_feature = torch.matmul(
+            q_feature.transpose(2, 3) / self.temperature, k_feature
+        )  # [batch_size, n_heads, d_v, d_v]
+
+        # apply masking on the attention map, this is optional
+        if attn_mask is not None:
+            attn_time = attn_time.masked_fill(attn_mask == 0, -1e9)
+
+        # compute attention score [0, 1], then apply dropout
+        attn_time = F.softmax(attn_time, dim=-1)
+        attn_feature = F.softmax(attn_feature, dim=-1)
+        if self.dropout is not None:
+            attn_time = self.dropout(attn_time)
+            attn_feature = self.dropout(attn_feature)
+
+        # multiply the score with v
+        output = torch.matmul(torch.matmul(attn_time, v), attn_feature)
+
+        return output, attn_time, attn_feature
+
+
+class MultiHeadCDSA(nn.Module):
+    """Multi-head Cross-Dimensional Self-Attention module.
+
+    Parameters
+    ----------
+    n_heads:
+        The number of heads in multi-head attention.
+
+    d_model:
+        The dimension of the input tensor.
+
+    d_k:
+        The dimension of the key and query tensor on time-series.
+
+    d_v:
+        The dimension of the value tensor,
+        The dimension of the key and query tensor on feature.
+
+    dropout:
+        The dropout rate.
+
+    attn_dropout:
+        The dropout rate for the attention map.
+
+    """
+
+    def __init__(
+        self,
+        n_heads: int,
+        d_model: int,
+        d_k: int,
+        d_v: int,
+        dropout: float,
+        attn_dropout: float,
+    ):
+        super().__init__()
+
+        self.n_heads = n_heads
+        self.d_k = d_k
+        self.d_v = d_v
+
+        # w_qs, w_ks for Time Dimension
```

Comment:
w_qs, w_ks for the time dimension and w_qs, w_ks for the feature dimension are not used in the func forward().